### ISSUE TITLE
Add SHA-2 (sha256 / sha512) to checksum

### ISSUE THREAD

**Issue by TomOne** [chocolatey/choco#113]
Although MD5 and SHA1 are still very common as checksums for files, both of these hash functions are considered insecure in cryptography. This is epecially true when these algorithms are used to hash passwords, which is a gross negligence.

When used as file checksums, an attacker could compromise a server and put a malicious file with the very same MD5 or SHA1 checksum as the original file onto that server. Although I know no real world case such an attack, it is still theoretically possible.

The solution is simple: deprecate MD5 and SHA1 as checksums and encourage maintainers to use the more recent SHA256 or SHA512 algorithms. Perhaps Chocolatey should give a warning message if one of the deprecated hash algorithms is used.

MD5 and SHA1 have really no place in modern cryptography and therefore it’s better to also not use them as file checksums.

**Comment by ferventcoder** [comment]
Blocked by at least https://github.com/ferventcoder/checksum/issues/1

**Comment by ferventcoder** [comment]
@TomOne Agreed. But it's not my option to deprecate the use of those checksums. That is going to be a change over time coming from the official sources that publish those hashes in many cases. We can start pushing on the world of those that offer the downloads to be more secure, but we should not fully deprecate until those locations offer it. SHA1, while still having the ability for file collisions is tremendously more secure than MD5. In some cases where the sites are HTTPS, someone would need to break into the site AND put a malicious file that matches checksum. Hard, but not impossible. 

For packages where the user is calculating those checksums (i.e. not offered by the site), we should definitely enforce a more secure checksum as a requirement.

**Comment by TomOne** [comment]
You’re right. I haven’t tought about that this further. :smile:

If we ever plan an (at least partially) automated moderation (what I would very much welcome), it should definitely check if MD5/SHA1 was used and give a warning to the maintainer with the message to use a stronger hash algorithm if available and perhaps encourage the software authors to to switch to SHA256/SHA512.

But these hash algorithms still have a drawback: If an attacker breaks into a site, replaces the file with a malicious file including its checksum **before** the corresponding Choco package is created, the checksum becomes useless for security purposes. GPG signatures wouldn’t have this problem, because to replace that file on the server with malicious code, an attacker would have to steal the private key of the signer, which is normally harder than breaking into the server. :smiling_imp: 

Please consider supporting GPG signatures too and make that the first choice for maintainers if available. I can remember that once a Chocolatey user asked for that, but I cannot find an issue on GH about it. Can I create an issue about that?

Even better would be to integrate GPG signatures for Chocolatey packages itself. Other package managers in the *nix world do that already.

**Comment by ferventcoder** [comment]
Actually plan to set GPG signatures on the packages themselves. That's part of the keys for the moderators conversation we've talked about.

**Comment by TomOne** [comment]
> Actually plan to set GPG signatures on the packages themselves.

Excellent, but actually we need support for both package signing and for the ability to provide a signature for the downloaded files. Package signing on it’s own doesn’t increase the security for externally downloaded files.

**Comment by ferventcoder** [comment]
Agreed. Yup. Best of both worlds is ultimately for packages to have no external downloads.

**Comment by TomOne** [comment]
> Agreed. Yup. Best of both worlds is ultimately for packages to have no external downloads.

Sure, but I doubt if that will ever happen, especially with proprietary software. If Chocolatey establishes as de-facto package manager for Windows, it is about to jeopardize the business model of adware authors and ads on download sites for Windows software. :smiling_imp: 

But back to the topic: To keep track of this topic, we should create issues for adding GPG signatures for packages and for external downloads, shouldn’t we?

**Comment by rismoney** [comment]
Packages without external downloads?  Binary stored where?

**Comment by TomOne** [comment]
> Packages without external downloads? Binary stored where?

In the package itself, otherwise it would again external.

**Comment by ferventcoder** [comment]
> But back to the topic: To keep track of this topic, we should create issues for adding GPG signatures for packages and for external downloads, shouldn’t we?

Yes, for packages. How would you create one for external downloads for applications you don't control?

**Comment by rismoney** [comment]
> In  the package itself, otherwise it would still be external.

this breaks version control.  This reminds me, we need a url environment variable for package portability.

**Comment by gep13** [comment]
> @TomOne said...
In the package itself, otherwise it would still be external.

> @rismoney said...
this breaks version control.

@ferventcoder am I not right in saying that going forward, the intention would be to offer a CDN style service for packages, so that we (Chocolatey) grab the MSI/EXE/Zip and store it "somewhere" then re-write the package download URL to point at the cached version.  Granted, a premium offering, but something that is being looked into, as per, kickstarter.

**Comment by ferventcoder** [comment]
> we need a url environment variable for package portability. 

Yup, still have it in mind. :)

**Comment by ferventcoder** [comment]
> @ferventcoder am I not right in saying that going forward, the intention would be to offer a CDN style service for packages, so that we (Chocolatey) grab the MSI/EXE/Zip and store it "somewhere" then re-write the package download URL to point at the cached version. Granted, a premium offering, but something that is being looked into, as per, kickstarter.

Yes, a premium offering.

**Comment by gep13** [comment]
> @ferventcoder said...
Yup, still have it in mind. :)

What is being referred to here?  Not sure I follow.

**Comment by ferventcoder** [comment]
> What is being referred to here? Not sure I follow.

That was mentioned there. @rismoney had a question.

**Comment by gep13** [comment]
> @ferventcoder said...
That was mentioned there. @rismoney had a question statement and I was agreeing.

What I was getting at is, what is, what is a `a url environment variable for package portability` and why is it required?  Don't recall seeing mention of this anywhere.

**Comment by ferventcoder** [comment]
We should open a new ticket about this - allows the package download files to be overridden.

**Comment by rednoah** [comment]
Is SHA-256 supported as `Install-ChocolateyPackage` `-checksumType` now? I see some code regarding SHA256 but it's not in the *Create Package* docs yet.

**Comment by ferventcoder** [comment]
@rednoah SHA2 is not in there yet. Only sha1 and md5 at the moment.

**Comment by StefanScherer** [comment]
I have played with `Get-FileHash node.msi -Algorithm sha256` to get some similar behavior as in Linux Docker Images for Windows PowerShell in this [Dockerfile](https://github.com/StefanScherer/dockerfiles-windows/blob/master/node/4.1/Dockerfile#L9)
But I don't know if this depends on a newer PowerShell that comes with Windows Server 2016 TP3.

**Comment by ferventcoder** [comment]
First version was v4 - https://technet.microsoft.com/en-us/library/dn520872(v=wps.630).aspx
